packages/teleport/src/Discover/Kubernetes/LoginTrait/LoginTrait.tsx

```diff
@@ -0,0 +1,288 @@
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { Text, Box, Indicator } from 'design';
+import * as Icons from 'design/Icon';
+
+import useTeleport from 'teleport/useTeleport';
+import {
+  SelectCreatable,
+  Option,
+} from 'teleport/Discover/Shared/SelectCreatable';
+import { AccessInfo } from 'teleport/Discover/Shared/AccessInfo';
+
+import {
+  Header,
+  HeaderSubtitle,
+  ActionButtons,
+  ButtonBlueText,
+} from '../../Shared';
+
+import { useLoginTrait, State } from './useLoginTrait';
+
+import type { AgentStepProps } from '../../types';
+
+const resourceName = 'Kubernetes';
+const traitDescription = 'users and groups';
+
+export default function Container(props: AgentStepProps) {
+  const ctx = useTeleport();
+  const state = useLoginTrait({ ctx, props });
+
+  return <LoginTrait {...state} />;
+}
+
+export function LoginTrait({
+  attempt,
+  nextStep,
+  dynamicGroups,
+  staticGroups,
+  dynamicUsers,
+  staticUsers,
+  fetchLoginTraits,
+  canEditUser,
+  isSsoUser,
+}: State) {
+  const [groupInputValue, setGroupInputValue] = useState('');
+  const [groups, setGroups] = useState<Option[]>(() => {
+    const fixedGroups = staticGroups.map(l => ({
+      value: l,
+      label: l,
+      isFixed: true,
+    }));
+    const groups = dynamicGroups.map(l => ({
+      value: l,
+      label: l,
+    }));
+    return [...fixedGroups, ...groups];
+  });
+
+  const [userInputValue, setUserInputValue] = useState('');
+  const [users, setUsers] = useState<Option[]>(() => {
+    const fixedUsers = staticUsers.map(l => ({
+      value: l,
+      label: l,
+      isFixed: true,
+    }));
+    const users = dynamicUsers.map(l => ({
+      value: l,
+      label: l,
+    }));
+    return [...fixedUsers, ...users];
+  });
+
+  const hasUsersOrGroups = groups.length > 0 || users.length > 0;
+  const canAddLoginTraits = !isSsoUser && canEditUser;
+
+  function onProceed() {
+    // Take out duplications and static users and groups.
+    const newDynamicUsers = new Set<string>();
+    users.forEach(o => {
+      if (!staticUsers.includes(o.value)) {
+        newDynamicUsers.add(o.value);
+      }
+    });
+
+    const newDynamicGroups = new Set<string>();
+    users.forEach(o => {
+      if (!staticGroups.includes(o.value)) {
+        newDynamicGroups.add(o.value);
+      }
+    });
+
+    nextStep([...newDynamicUsers], [...newDynamicGroups]);
+  }
+
+  function handleGroupKeyDown(event) {
+    if (!groupInputValue) return;
+    switch (event.key) {
+      case 'Enter':
+      case 'Tab':
+        setGroups([
+          ...groups,
+          { value: groupInputValue, label: groupInputValue },
+        ]);
+        setGroupInputValue('');
+        event.preventDefault();
+    }
+  }
+
+  function handleUserKeyDown(event) {
+    if (!userInputValue) return;
+    switch (event.key) {
+      case 'Enter':
+      case 'Tab':
+        setUsers([...users, { value: userInputValue, label: userInputValue }]);
```

Comment:
```suggestion
setUsers(users => [...users, { value: userInputValue, label: userInputValue }]);
```

Reply:
hm, what is the difference? is this ensuring we use the latest `users`?

Reply:
When computing the next state based on prev state is IMO better to use this approach. We are not currently enforcing  `react-hooks/exhaustive-deps` so it's relatively easy to introduce stale closure that is used as an event handler/in `useEffect`. Using `(prev) => ...` we eliminate some of this. See the example https://codesandbox.io/s/awesome-mcnulty-wdwdky?file=/src/App.tsx.